### DISCUSSION TITLE
Status code Retry

### DISCUSSION THREAD

**Discussion by liukaiyuliukaiyu** [Kong/kong#12326] [Help]
How does Kong support custom status code retry functionality, such as retry when the service returns a 500 status code
The proxy_next_upstream configuration is available only globally

**Comment by NPC403** [comment]
In Kong, you can achieve custom status code retry functionality, such as retrying when a service returns a 500 status code, using the `proxy_next_upstream` and `upstream_retry` directives in your Kong configuration. While the `proxy_next_upstream` directive is indeed global, you can use it in combination with the `upstream_retry` directive to achieve more granular retry behavior based on status codes.

Here's an example of how you can configure Kong to retry requests when the upstream service returns a 500 status code:

```nginx
upstream my_upstream {
    server backend1.example.com;
    server backend2.example.com;
    server backend3.example.com;
}

server {
    location / {
        proxy_pass http://my_upstream;
        proxy_next_upstream error timeout http_500;  # Retry on 500 status code
        upstream_retry max=5 timeout=1s;
    }
}
```
In this example:

- `proxy_next_upstream error timeout http_500;` specifies that Kong should retry the request if the upstream server returns an error, a timeout, or a 500 status code.
- `upstream_retry max=5 timeout=1s;` sets up the retry mechanism with a maximum of 5 retries and a timeout of 1 second between retries.

By combining these directives in your Kong configuration, you can customize the retry behavior based on specific status codes, such as retrying only on 500 status codes, while still having control over the retry count and timeout.


Answered by ChatGPT

**Reply by liukaiyuliukaiyu** [reply]
I know how to use this configuration proxy_next_upstream. After using this configuration, all the services that access the gateway will take effect this configuration. I want to realize that the services that configure the 500 status code retry will take effect, and those that do not configure will not take effect

**Comment by tzssangglass** [comment]
AFAIK, it can not apply a specific retry strategy for a service

**Reply by liukaiyuliukaiyu** [reply]
Is it possible to use lua plugins to do this ？not use nginx proxy_next_upstream

**Reply by tzssangglass** [reply]
may not
